### PR TITLE
Add simple MLIR SLP vectorizer

### PR DESCRIPTION
**Pull request by Hardcode84**
Based on upstream PR (https://github.com/llvm/llvm-project/pull/140469)

Simple SLP vectorizer to combine sequence of scalar ops and smaller vector ops into larger vector ops. Doesn't have any cost model (yet?) therefore "greedy". See code comments for impl details. Mostly adding it to test it in Wave but if you have a good place to plug it in IREE itself we can test it there as well.

### CODE REVIEW

**Comment by benvanik** [comment]
Let's get feedback from the codegen team - if we don't think we'll be using this we'll not want to land it. /cc @MaheshRavishankar (when you're back)

File: compiler/src/iree/compiler/Codegen/Dialect/VectorExt/Transforms/Passes.td [hunk 1]
```diff
@@ -28,4 +28,24 @@ def VectorExtFoldUnitExtentDimsPass :
   ];
 }
 
+def GreedySLPVectorizerPass : Pass<"iree-greedy-slp-vectorizer"> {
```
Comment (benvanik) [inline compiler/src/iree/compiler/Codegen/Dialect/VectorExt/Transforms/Passes.td:31]:
```suggestion
def GreedySLPVectorizerPass : Pass<"iree-vector-ext-greedy-slp-vectorizer"> {
```

File: compiler/src/iree/compiler/Codegen/Dialect/VectorExt/Transforms/test/BUILD.bazel [hunk 2]
```diff
@@ -18,6 +18,7 @@ iree_lit_test_suite(
     name = "lit",
     srcs = enforce_glob(
         [
+            "slp-vectorize.mlir",
```
Comment (benvanik) [inline compiler/src/iree/compiler/Codegen/Dialect/VectorExt/Transforms/test/BUILD.bazel:21]:
```suggestion
            "slp_vectorize.mlir",
```

File: compiler/src/iree/compiler/Codegen/Dialect/VectorExt/Transforms/test/slp-vectorize.mlir [hunk 3]
```diff
@@ -0,0 +1,913 @@
+// RUN: iree-opt %s -pass-pipeline='builtin.module(func.func(iree-greedy-slp-vectorizer{max-vector-bitwidth=256}))' | FileCheck %s
+
+
```
Comment (benvanik) [inline compiler/src/iree/compiler/Codegen/Dialect/VectorExt/Transforms/test/slp-vectorize.mlir:3]:
```suggestion

```

File: compiler/src/iree/compiler/Codegen/Dialect/VectorExt/Transforms/test/slp-vectorize.mlir [hunk 4]
```diff
@@ -0,0 +1,913 @@
...
+  %0 = memref.load %arg0[%c0] : memref<8xi32>
+  %4 = memref.load %arg1[%c0] : memref<8xi32>
+  %8 = arith.addi %0, %4 : i32
+  memref.store %8, %arg0[%c0] : memref<8xi32>
+
+  return
+}
+
+
```
Comment (benvanik) [inline compiler/src/iree/compiler/Codegen/Dialect/VectorExt/Transforms/test/slp-vectorize.mlir:17]: add splits to this file as with other tests and remove extraneous whitespace

```suggestion

// -----
```

File: compiler/src/iree/compiler/Codegen/Dialect/VectorExt/Transforms/test/slp-vectorize.mlir [hunk 5]
```diff
@@ -0,0 +1,913 @@
+// RUN: iree-opt %s -pass-pipeline='builtin.module(func.func(iree-greedy-slp-vectorizer{max-vector-bitwidth=256}))' | FileCheck %s
```
Comment (benvanik) [inline compiler/src/iree/compiler/Codegen/Dialect/VectorExt/Transforms/test/slp-vectorize.mlir:1]:
```suggestion
// RUN: iree-opt %s --split-input-file --pass-pipeline='builtin.module(func.func(iree-greedy-slp-vectorizer{max-vector-bitwidth=256}))' | FileCheck %s
```

File: compiler/src/iree/compiler/Codegen/Dialect/VectorExt/Transforms/test/slp-vectorize.mlir [hunk 6]
```diff
@@ -0,0 +1,913 @@
...
+  %11 = arith.addi %3, %7 : i32
+
+  return %8, %9, %10, %11, %12, %13, %14, %15 : i32, i32, i32, i32, i32, i32, i32, i32
+}
+
+
+// CHECK-LABEL: func @read_read_add_add_vec
+//  CHECK-SAME: (%[[ARG0:.*]]: memref<8xi32>, %[[ARG1:.*]]: memref<8xi32>)
+func.func @read_read_add_add_vec(%arg0: memref<8xi32>, %arg1: memref<8xi32>) ->
+                                   (vector<2xi32>, vector<2xi32>){
```
Comment (benvanik) [inline compiler/src/iree/compiler/Codegen/Dialect/VectorExt/Transforms/test/slp-vectorize.mlir:678]:
```suggestion
                                   (vector<2xi32>, vector<2xi32>) {
```
here and elsewhere

File: compiler/src/iree/compiler/Codegen/Dialect/VectorExt/Transforms/SLPVectorizer.cpp [hunk 7]
```diff
@@ -0,0 +1,1262 @@
...
+  IRRewriter rewriter(block.getParentOp()->getContext());
+  FailureOr<size_t> numNodesVectorized =
+      graph.vectorize(rewriter, isValidVecType);
+  if (failed(numNodesVectorized))
+    LLVM_DEBUG(llvm::dbgs() << "Failed to vectorize graph\n");
+
+  return numNodesVectorized;
+}
+
+static bool isPow2(size_t size) {
```
Comment (benvanik) [inline compiler/src/iree/compiler/Codegen/Dialect/VectorExt/Transforms/SLPVectorizer.cpp:1216]: you can use llvm::isPowerOf2

File: compiler/src/iree/compiler/Codegen/Dialect/VectorExt/Transforms/SLPVectorizer.cpp [hunk 8]
```diff
@@ -0,0 +1,1262 @@
...
+  // Build the SLP graph from root groups
+  SLPGraph graph = buildSLPGraph(rootGroups);
+  LLVM_DEBUG(graph.print());
+
+  // Vectorize the graph
+  IRRewriter rewriter(block.getParentOp()->getContext());
+  FailureOr<size_t> numNodesVectorized =
+      graph.vectorize(rewriter, isValidVecType);
+  if (failed(numNodesVectorized))
+    LLVM_DEBUG(llvm::dbgs() << "Failed to vectorize graph\n");
```
Comment (benvanik) [inline compiler/src/iree/compiler/Codegen/Dialect/VectorExt/Transforms/SLPVectorizer.cpp:1211]: use braces when including macros or multi-line statements - it's really easy to have this go wrong

but when doing logic like this you want to put the whole thing in a block - otherwise you have an if with no body and it makes it more difficult to read (a user doesn't know if this is missing a failure return and the logic is guarding something or of it's just debug-related)
```suggestion
  LLVM_DEBUG({
    if (failed(numNodesVectorized))
      llvm::dbgs() << "Failed to vectorize graph\n");
  });
```

File: compiler/src/iree/compiler/Codegen/Dialect/VectorExt/Transforms/SLPVectorizer.cpp [hunk 9]
```diff
@@ -0,0 +1,1262 @@
...
+  // Process each group to find contiguous sequences
+  SmallVector<MemoryOpGroup> rootGroups;
+  for (const auto &group : groups) {
+    SmallVector<MemoryOpGroup> contiguousGroups =
+        extractContiguousGroups(group);
+    LLVM_DEBUG({
+      llvm::dbgs() << "Found " << contiguousGroups.size()
+                   << " contiguous groups in "
+                   << (group.isLoadGroup() ? "load" : "store") << " group\n";
+      for (const auto &contigGroup : contiguousGroups) {
```
Comment (benvanik) [inline compiler/src/iree/compiler/Codegen/Dialect/VectorExt/Transforms/SLPVectorizer.cpp:1194]: keep names consistent
```suggestion
      for (const auto &contiguousGroup : contiguousGroups) {
```

**Review by benvanik — CHANGES_REQUESTED** [review]
As mentioned let's make sure this is something IREE codegen wants. If it's not you'll want to keep it downstream - we generally don't want to have experimental/speculative code checked unless it's on a production path.

**Comment by qedawkins** [comment]
This warrants a VC for sure. The core idea with structured op codegen is to *not* have to do this. I could see a world where we don't mind adding this, but it'll be really important to avoid relying on it as a crutch.

**Comment by Hardcode84** [comment]
(I didn't even undraft it yet)
